dbms/src/Storages/Page/V3/PageEntryCheckpointInfo.h

```diff
@@ -46,10 +47,58 @@ struct CheckpointLocation
         CheckpointProto::StringsInternMap & strings_map);
 };
 
-struct CheckpointInfo
+// A more memory compact struct compared to std::optional<CheckpointInfo>
+class OptionalCheckpointInfo
 {
+public:
+    // ctor
+    OptionalCheckpointInfo(std::nullopt_t) // NOLINT(google-explicit-constructor)
+        : is_set(false)
+    {}
+    OptionalCheckpointInfo(CheckpointLocation data_loc, bool is_local_data_reclaimed_)
+        : data_location(std::move(data_loc))
+        , is_set(true)
+        , is_local_data_reclaimed(is_local_data_reclaimed_)
+    {
+    }
+
+    // Act as if std::optional<...>
+    ALWAYS_INLINE bool has_value() const { return is_set; } // NOLINT(readability-identifier-naming)
+    ALWAYS_INLINE OptionalCheckpointInfo * operator->()
+    {
+        assert(is_set); // should only be used when contains valid value
+        return this;
+    }
+    ALWAYS_INLINE const OptionalCheckpointInfo * operator->() const
+    {
+        assert(is_set); // should only be used when contains valid value
+        return this;
+    }
```

Comment:
Seems that this will not prevent us to use it like `checkpoint_info.data_location`?

Also these constructors prevented us from use it in a named way (we must construct its members using unnamed parameters).

How about simplify the usage as the follows, keeping it as a simple POD struct:

```c++
auto x = CheckpointInfo{}; // Empty Checkpoint

auto x = CheckpointInfo{
    .data_location = xxx,
    .is_local_data_reclaimed = xxx,
    .is_valid = true,
}; // Non Empty checkpoint
```